src/client/mod.rs

```diff
@@ -169,11 +169,14 @@ impl APIClient {
                             // append it to our current buffer
                             buff.extend_from_slice(&chunk);
 
-                            // If it ends with a newline, then it should be a complete resource
-                            // so we try to parse it.
-                            // On success we yield this item
-                            // otherwise we either continue looping or break and return an error
-                            if buff.ends_with(b"\n") {
+                            // Check to see if we have valid JSON in our buffer,
+                            // and if so parse it fully and return.
+                            // Deserializing as serde::de::IgnoredAny is just a syntax check
+                            // and does not go through any of the memory intensity of constructing the actual
+                            // value.
+                            let valid = serde_json::from_slice::<IgnoredAny>(&buff);
```

Comment:
i still think this won't necessarily be enough because `buff` can be more than one complete object. If we're following how k8s flushes, `buff` be two complete objects, it can be 1.5.

we probably need to do a `for item in buff.lines()`, parse every complete line, and store eof remains in the buffer for the outer loop.

Reply:
Okay that makes total sense, thanks for that. I've updated the unfold to that end, it's gotten a good bit uglier but in my testing it's been working. I've also been testing with the `configmap_reflector` example using a cm full of new lines, and those have been parsed correctly.

I found that I needed to use a third buffer of parsed items in the unfold, as it was very finicky pulling out just the first complete line from `buff` and then saving the rest for the next iteration, though this has forced me to add a `K: Send` to some methods, which doesn't seem too unreasonable, but worth pointing out.

Reply:
Nice work. It's certainly getting into the hairy territory now, but we're almost there.

Posted some potential clean-ups on the pr, then there's the other issue; I am able to crash this now with the `node_informer` by just running for 1 minute locally;

```
[2019-11-25T19:10:35Z DEBUG node_informer] Normal node: ip-10-100-38-XX.eu-west-2.compute.internal
Error: Error deserializing response

Caused by:
    invalid type: integer `1`, expected adjacently tagged enum WatchEvent at line 1 column 1
`

Reply:
Another crash in node_informer when it tried to parse an incomplete chunk (missing start):

[2019-11-25T19:35:27Z WARN  kube::client] Parsing: ames":["quay.io/cluxblah@sha256:fbd9d57d5bfc08fa10ca6e32e5095355ac4256e0ff5413d0984d983a5a47b3f4",...remains of valid object that ends correct...

I would suggest the following for people reporting errors:

```diff
diff --git a/src/client/mod.rs b/src/client/mod.rs
index f5560a9c..7f062116 100644
--- a/src/client/mod.rs
+++ b/src/client/mod.rs
@@ -207,8 +207,11 @@ impl APIClient {
                                                 // and add the newline character we stripped as part of the iterator.
                                                 new_buff.extend_from_slice(&line);
                                                 new_buff.push(b'\n');
-                                            }
-                                            Err(e) => items.push(Err(Error::SerdeError(e))),
+                                            },
+                                            Err(e) => {
+                                                warn!("Parsing: {}", String::from_utf8_lossy(line));
+                                                items.push(Err(Error::SerdeError(e)));
+                                            },
                                         }
                                     }
 
```

Reply:
I've pushed some changes that remove the need for `Send` and clean up the `unfold` quite a bit, though I'm still testing the crashes you've mentioned here. I haven't been able to reproduce them, I need to get on a chattier kube api server 🤷‍♂

Reply:
Nice. I'm not seeing the crash anymore. At least it's survived half an hour. You might have accidentally fixed it.

Reply:
That's the best kind of fix 🎆